docs/proposals/04-new-core-gardener-cloud-apis.md

```diff
@@ -0,0 +1,718 @@
+    #     syncPeriod: 30s
+    #     tolerance: 0.1
+    #-#-# only usable with Kubernetes < 1.12
+    #     downscaleDelay: 15m0s
+    #     upscaleDelay: 1m0s
+    #-#-# only usable with Kubernetes >= 1.12
+    #     downscaleStabilization: 5m0s
+    #     initialReadinessDelay: 30s
+    #     cpuInitializationPeriod: 5m0s
+    # kubeScheduler:
+    #   featureGates:
+    #     SomeKubernetesFeature: true
+    # clusterAutoscaler:
+    #   scaleDownUtilizationThreshold: 0.5
+    #   scaleDownUnneededTime: 30m
+    #   scaleDownDelayAfterAdd: 60m
+    #   scaleDownDelayAfterFailure: 10m
+    #   scaleDownDelayAfterDelete: 10s
+    #   scanInterval: 10s
+    worker:
+    - name: cpu-worker
+      minimum: 3
+      maximum: 5
+    # maxSurge: 1
+    # maxUnavailable: 0
+      machine:
+        type: m5.large
+        image:
+          name: <some-os-name>
+          version: <some-os-version>
+        # providerConfig:
+        #   <some-os-specific-configuration>
+      volume:
+        type: gp2
+        size: 20Gi
+    # providerConfig:
+    #   <some-provider-specific-worker-config>
+    # labels:
+    #   key: value
+    # annotations:
+    #   key: value
+    # taints: # See also https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/
+    # - key: foo
+    #   value: bar
+    #   effect: NoSchedule
+    # kubernetes:
+    #   kubeProxy:
+    #     featureGates:
+    #       SomeKubernetesFeature: true
+    #     mode: IPVS
+    #   kubelet:
+    #     cpuCFSQuota: true
+    #     cpuManagerPolicy: none
+    #     podPidsLimit: 10
+    #     featureGates:
+    #       SomeKubernetesFeature: true
+      zones:
+      - europe-central-1a
+      - europe-central-1b
+  kubernetes:
+    version: 1.15.1
+  # allowPrivilegedContainers: true # 'true' means that all authenticated users can use the "gardener.privileged" PodSecurityPolicy, allowing full unrestricted access to Pod features.
+  dns:
+    # When the shoot shall use a cluster domain no domain and no providers need to be provided - Gardener will
+    # automatically compute a correct domain.
+    domain: crazy-botany.core.my-custom-domain.com
+    providers:
+    - type: aws-route53
+      secretName: my-custom-domain-secret
+      domains:
+        include:
+        - my-custom-domain.com
+        - my-other-custom-domain.com
+        exclude:
+        - yet-another-custom-domain.com
+  extensions:
+  - type: foobar
+  # providerConfig:
+  #   apiVersion: foobar.extensions.gardener.cloud/v1alpha1
+  #   kind: FooBarConfiguration
+  #   foo: bar
+  networking:
+    type: calico
+    pods: 100.96.0.0/11
+    services: 100.64.0.0/13
+    nodes: 10.250.0.0/16
+  # providerConfig:
+  #   apiVersion: calico.extensions.gardener.cloud/v1alpha1
+  #   kind: NetworkConfig
+  #   ipam:
+  #     type: host-local
+  #     cidr: usePodCIDR
+  #   backend: bird
+  # See also: https://github.com/gardener/gardener/blob/master/docs/proposals/03-networking.md
+  maintenance:
```

Comment:
IMHO I would rather move this field in a separate resource:

```yaml
apiVersion: core.gardener.cloud/v1alpha1
kind: MaintenanceTarget
metadata:
  name: only-during-lunch
  namespace: some-namespace
spec:
  shootName: my-shoot-cluster
  timeWindow:
    begin: 110000+0100
    end: 120000+0100
  autoUpdate:
    kubernetesVersion: true
    machineImageVersion: true
```

Then a controller is going to modify a single boolean called `reconcile` in the `Shoot` specification and update the K8S versions and/or machine versions.

The beauty of this approach is that it can be used to create higher level abstractions such as

```yaml
apiVersion: core.gardener.cloud/v1alpha1
kind: Maintenance 
metadata:
  name: only-during-night
  namespace: some-namespace
spec:
  shootSelector:
    matchLabels:
      schedule: at-night
  selector:
    matchLabels:
      schedule: at-night
  template:
    metadata:
      labels:
        schedule: at-night
    spec:
      timeWindow:
        begin: 220000+0100
        end: 230000+0100
      autoUpdate:
        kubernetesVersion: true
        machineImageVersion: false
```

Which uses a shoot selector and then it creates a `MaintenanceTarget` for each shoot:

```yaml
apiVersion: core.gardener.cloud/v1alpha1
kind: MaintenanceTarget
metadata:
  name: only-during-night-4234t
  namespace: some-namespace
spec:
  shootName: some-shoot
  timeWindow:
    begin: 220000+0100
    end: 230000+0100
  autoUpdate:
    kubernetesVersion: true
    machineImageVersion: false
```

Reply:
I'm not convinced... It complicates everything and makes it less understandable for end-users, because suddenly they have to interact with multiple resources which somehow belong to a shoot. Can you highlight the benefits? Which problems would solve such approach?

> The beauty of this approach is that it can be used to create higher level abstractions such as

I don't understand this "beauty", sorry, can you explain? How do these higher level abstractions help?

If you like something like suggested in #1161 could be done after you introduced the new `settings.gardener.cloud` API group - but for maintenance I also don't see a strong need for that.

Reply:
> It complicates everything and makes it less understandable for end-users, because suddenly they have to interact with multiple resources which somehow belong to a shoot. Can you highlight the benefits? Which problems would solve such approach?

Clear separation of concerns - the `Shoot` represents the desired state of a cluster. Period.
When and what updates the Shoot fields (image/kube versions) is another optional concern which should be put in a separate resource, because it interacts with fields declared in the `Shoot` resource.

Reply:
Sorry, I can't follow these artificial arguments right now... I still think it complicates the life of end-users, and I don't see a significant benefit in doing this.
Anyway, we should definitely not do it now as we want to focus on completing #308, but even if you want to do it later, maybe it's a good idea to gather some end-user feedback on this one first. I think end-users will find this overcomplicated and hard to use.